plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,324 @@
+  password = "pass123" # required
+  ## XtremIO User Interface Endpoint
+  url = "https://xtremio.example.com/" # required
+  ## Metrics to collect from the XtremIO
+  collectors = ["bbus","clusters","ssds","volumes","xms"]
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+// Description will appear directly above the plugin definition in the config file
+func (xio *XtremIO) Description() string {
+	return `Gathers Metrics From a Dell EMC XtremIO Storage Array's V3 API`
+}
+
+// SampleConfig will populate the sample configuration portion of the plugin's configuration
+func (xio *XtremIO) SampleConfig() string {
+	return sampleConfig
+}
+
+func (xio *XtremIO) Init() error {
+	if xio.Username == "" {
+		return fmt.Errorf("Username cannot be empty")
+	}
+	if xio.Password == "" {
+		return fmt.Errorf("Password cannot be empty")
+	}
+	if xio.URL == "" {
+		return fmt.Errorf("URL cannot be empty")
+	}
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+	err = xio.Authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	availableCollectors := []string{"bbus", "clusters", "ssds", "volumes", "xms"}
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+			if !contains(availableCollectors, collector) {
+				acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				return
+			}
+
+			resp, err := xio.Call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			} else {
+				arr = gjson.Get(resp, collector).Array()
+			}
+
+			for _, item := range arr {
+				itemSplit := strings.Split(gjson.Get(item.Raw, "href").Str, "/")
+				url := ""
+				if len(itemSplit) > 0 {
+					url = collector + "/" + itemSplit[len(itemSplit)-1]
+				} else {
+					continue
+				}
+
+				if collector == "bbus" {
+					wg.Add(1)
+					go xio.GatherBBUs(&wg, url, acc)
+				}
+				if collector == "clusters" {
+					wg.Add(1)
+					go xio.GatherClusters(&wg, url, acc)
+				}
+				if collector == "ssds" {
+					wg.Add(1)
+					go xio.GatherSSDs(&wg, url, acc)
+				}
+				if collector == "volumes" {
+					wg.Add(1)
+					go xio.GatherVolumes(&wg, url, acc)
+				}
+				if collector == "xms" {
+					wg.Add(1)
+					go xio.GatherXMS(&wg, url, acc)
+				}
+			}
+		}(collector)
+	}
+
+	wg.Wait()
+
+	xio.ResetCookie()
+
+	return nil
+}
+
+func (xio *XtremIO) GatherBBUs(wg *sync.WaitGroup, url string, acc telegraf.Accumulator) {
```

Comment:
Do not export those functions please and _always_ make the accumulator the first parameter.
```suggestion
func (xio *XtremIO) gatherBBUs(acc telegraf.Accumulator, url string) {
```
With the removal of the parallelization you don't need to pass in the wait-group anymore.

Reply:
Leaving parallelization in for now, so wait group is still being passed.